src/ImageSharp/Formats/Png/PngDecoderCore.cs

```diff
@@ -924,10 +919,44 @@ private void ProcessInterlacedDefilteredScanline<TPixel>(ReadOnlySpan<byte> defi
         }
     }
 
+    /// <summary>
+    /// Decodes and assigns the color palette to the metadata
+    /// </summary>
+    /// <param name="palette">The palette buffer.</param>
+    /// <param name="alpha">The alpha palette buffer.</param>
+    /// <param name="pngMetadata">The png metadata.</param>
+    private static void AssignColorPalette(ReadOnlySpan<byte> palette, ReadOnlySpan<byte> alpha, PngMetadata pngMetadata)
+    {
+        if (palette.Length == 0)
+        {
+            return;
+        }
+
+        Color[] colorTable = new Color[palette.Length / Unsafe.SizeOf<Rgb24>()];
+        ref Rgb24 paletteBase = ref MemoryMarshal.GetReference(MemoryMarshal.Cast<byte, Rgb24>(palette));
+        for (int i = 0; i < colorTable.Length; i++)
+        {
+            colorTable[i] = new Color(Unsafe.Add(ref paletteBase, (uint)i));
+        }
+
+        if (alpha.Length > 0)
+        {
+            // The alpha chunk may contain as many transparency entries as there are palette entries
+            // (more than that would not make any sense) or as few as one.
+            for (int i = 0; i < alpha.Length; i++)
+            {
+                ref Color color = ref colorTable[i];
+                color = color.WithAlpha(alpha[i] / 255F);
```

Comment:
If perf is important here, then
```suggestion
                color = color.WithAlpha(alpha[i] * (1F / 255F));
```
will produce a [mul instead of div](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGAzADYRsGBgFlyACgCWAO1HYAlAwC8APgbYGAegakArPr4BuLj3oNBw0WNIz5m5es0MAVA0nk+OvYb6LTagBfIA==).

Even if perf isn't critical here I think that I'd change the code, as
* it's almost as readable
* it's not only about the perf of this piece of code, but rather for the whole system, as the cpu-ports can be utilized better when there's no division clogging it

Note: it's Roslyn to blame here for not being able to emit `1 / 255F` automatically -- though the JIT could do that optimization too, but it has more stringent timing constraints...

Reply:
This would not be a valid optimization for either compiler to make, as the results are not the same (due to 1/255f not being precisely representable in float).  See https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AGQEs8BHAbmqIDMjEvQDC9AN7V6M4SKlVZS+gDNoMAIZgAFvQAUnDPQBu9TvQCieHAFsYUDcAA2MZgCUNeAOYw9tNPQkAKxBAJShktLK0ewq+qYA9IEhcQCEALwm9ABU+mQJwUEq4VHRZYxkAJx6ACQARBLGAL4gkonJRU30GW05eQUpoU11oXyKZU2l9JNUTUA===

For that reason, unless it's a hot path, this optimization loses precision for no measurable benefit.

Reply:
> as the results are not the same

Yep, but the error is in the region of 1e-8 or less ([demo](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AGQEs8BHAbmqIDMjEvQDC9AN7V6M4SKlVZS+gDNoMAIZgAFvQAUnDPQBu9TvQCieHAFsYUDcAA2MZgCUNeAOYw9tNPQkAKxBAJShktLK0ewq+qYA9IEhcQCEALwm9ABU+mQJwUEq4VHRZYxkAJx6ACQARBLGAL4gkonJRU30GW05eQUpoU0AOhiN9EmFcXBZuXr5U0N1oXyKZU2l9BtUTUA===))

Reply:
In that value range, yes, but the magnitude of the error is relative to the magnitude of the value -- that's how floating point works.  Whether the precision loss is acceptable in any given case is a question of the trade-off between performance (~12 cycles for `divss` vs ~4 for `mulss`) and precision.

I'm not familiar enough with the code to know whether that trade-off is worthwhile here (though it looks questionable).  Mostly, I was commenting on your suggestion that Roslyn and RyuJIT were somehow missing an optimization one of them could have made automatically, which is incorrect.

Reply:
> which is incorrect

You are right (and I have to apologize here by Roslyn).
I had a look what GCC does (before writing the first comment on this) and it emitted the `mul` -- but I had `-Ofast` compiler switch set which disregards strict standards compliance (as I re-read just now). With `-O{2,3}` it's a `div`.

Thanks for the heads up!